src/irq.c

```diff
@@ -68,153 +68,167 @@ static int irq_config(const char *key, const char *value) {
+    if (strncmp(cpu_names[i], "CPU", 3) == 0) {
+      cpu_names[i] += 3;
+    }
+  }
+
+  char buffer[1024] = {0};
   while (fgets(buffer, sizeof(buffer), fh) != NULL) {
-    char *irq_name;
-    size_t irq_name_len;
-    derive_t irq_value;
-    int i;
-    int fields_num;
-    int irq_values_to_parse;
-
-    fields_num = strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
+    strstripnewline(buffer);
+
+    /* There is one column with the interrupt ID before the CPU counters and
+     * there may be up to three columns after the counters. */
+    char *fields[cpu_count + 8];
+    memset(fields, 0, sizeof(fields));
+
+    int fields_num = irq_strsplit(buffer, fields, STATIC_ARRAY_SIZE(fields));
     if (fields_num < 2)
       continue;
 
     /* Parse this many numeric fields, skip the rest
      * (+1 because first there is a name of irq in each line) */
-    if (fields_num >= cpu_count + 1)
-      irq_values_to_parse = cpu_count;
-    else
+    int irq_values_to_parse = cpu_count;
+    if (fields_num < cpu_count + 1) {
       irq_values_to_parse = fields_num - 1;
+    }
 
     /* First field is irq name and colon */
-    irq_name = fields[0];
-    irq_name_len = strlen(irq_name);
-    if (irq_name_len < 2)
+    char *irq_name = fields[0];
+    size_t irq_name_len = strlen(irq_name);
+    if ((irq_name_len < 2) || (irq_name[irq_name_len - 1] != ':')) {
       continue;
+    }
+    /* strip colon */
+    irq_name[irq_name_len - 1] = 0;
+    irq_name_len--;
 
-    /* Check if irq name ends with colon.
-     * Otherwise it's a header. */
-    if (irq_name[irq_name_len - 1] != ':')
+    /* Is it the the ARM fast interrupt (FIQ)? */
+    if (strcmp(irq_name, "FIQ") == 0)
       continue;
 
-    /* Is it the the ARM fast interrupt (FIQ)? */
-    if (irq_name_len == 4 && (strncmp(irq_name, "FIQ:", 4) == 0))
+    if (ignorelist_match(ignorelist, irq_name) != 0)
       continue;
 
-    irq_name[irq_name_len - 1] = '\0';
-    irq_name_len--;
+    metric_t m = {0};
+    metric_label_set(&m, "id", irq_name);
+
+    if (fields_num == cpu_count + 4) {
+      metric_label_set(&m, "device", fields[fields_num - 3]);
+      metric_label_set(&m, "trigger", fields[fields_num - 2]);
+      metric_label_set(&m, "kernel_module", fields[fields_num - 1]);
+    } else if (fields_num == cpu_count + 2) {
+      // in this case, the last column is a human-readable name
+      metric_label_set(&m, "name", fields[fields_num - 1]);
+    } else if (fields_num <= cpu_count + 1) {
+      // no-op
+    } else {
+      DEBUG("irq plugin: got %d fields, want %d or %d", fields_num,
+            fields_num + 4, fields_num + 2);
+    }
 
-    irq_value = 0;
-    for (i = 1; i <= irq_values_to_parse; i++) {
+    for (int i = 1; i <= irq_values_to_parse; i++) {
       /* Per-CPU value */
       value_t v;
-      int status;
-
-      status = parse_value(fields[i], &v, DS_TYPE_DERIVE);
+      int status = parse_value(fields[i], &v, DS_TYPE_COUNTER);
```

Comment:
DERIVE -> COUNTER?

Reply:
Yeah, don't know why this change was made. I think for version 6 we should move away from unsigned integers with overflow semantic. Reverted.

Reply:
Problem with metric types is counters being integers (regardless of whether they're unsigned or signed), because _Prometheus_ / _OpenMetrics_ wants base units to be used for reporting.

Reporting usage seconds accurately (to differentiate between complete idle vs. very little usage) would require counters to be floating point, but `collectd` does not support that, so sysman plugin needs to report microseconds instead...